### PR TITLE
Feature Support Header & Configurable Retry

### PR DESCRIPTION
**Pull request by leonardoventurini**
- [x] Receives and parses an `Allowed-Features` header.
- [x] Upgrades `axios`.
- [x] Makes Retry configurable from the agent.
- [x] Convert JSON String to Buffer before sending request.

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import retry, { ByPassRetryError } from './retry';
+import axios from 'axios';
+import debug from 'debug';
+import { HttpMethod } from './constants';
+
+const logger = debug('kadira-core:transport');
+
+export function promisifyStream(stream) {
```
Comment (leonardoventurini) [inline lib/utils.js:8]: Useful for comparing the timings, but not used in the code itself.

File: lib/index.js [hunk 2]
```diff
@@ -133,62 +138,47 @@ export class Kadira {
...
+        ...params.headers,
+      },
+    }).then(res => res.data);
+  }
 
-    if (params.noRetry) {
-      retryEnabled = false;
-      delete params.noRetry;
+  _sendJsonStream(url, params) {
+    if (hasCircularReference(params.data)) {
```
Comment (zodern) [inline lib/index.js:169]: Why is this step necessary? How does zipson-stream handle circular references?

File: lib/index.js [hunk 3]
```diff
@@ -81,26 +84,28 @@ export class Kadira {
   }
 
   // send the given payload to the server
-  sendData(_payload) {
+  async sendData(_payload) {
```
Comment (zodern) [inline lib/index.js:87]: This `async` doesn't seem to be needed.

File: lib/index.js (hunk 2 above)
Comment (leonardoventurini) [reply lib/index.js:169]: It gives a maximum call stack reached error, but is not interceptable by the catch it seems.

File: lib/index.js (hunk 3 above)
Comment (leonardoventurini) [reply lib/index.js:87]: I added it so it always returns a promise as is expected.

Comment (leonardoventurini) [reply lib/index.js:87]: Mostly due to the throw nested in it... it might work without, will check.

Comment (leonardoventurini) [reply lib/index.js]: It did, but I had to surround it with a Promise.

**Pull request closed without merging**